src/main/java/com/google/devtools/build/lib/bazel/repository/cache/RepositoryCache.java

```diff
@@ -276,19 +285,26 @@ private void ensureCacheDirectoryExists(KeyType keyType) throws IOException {
   /**
    * Assert that a file has an expected checksum.
    *
+   * @param xattrProvider The provider to use for extended attributes.
    * @param expectedChecksum The expected checksum of the file.
    * @param filePath The path to the file.
    * @param keyType The type of hash function. e.g. SHA-1, SHA-256
    * @throws IOException If the checksum does not match or the file cannot be hashed, an exception
    *     is thrown.
    */
-  public static void assertFileChecksum(String expectedChecksum, Path filePath, KeyType keyType)
+  public static void assertFileChecksum(XattrProvider xattrProvider, String expectedChecksum, Path filePath, KeyType keyType)
       throws IOException, InterruptedException {
     Preconditions.checkArgument(!expectedChecksum.isEmpty());
 
     String actualChecksum;
     try {
-      actualChecksum = getChecksum(keyType, filePath);
+      if (xattrProvider != null) {
+        actualChecksum = BaseEncoding.base16().lowerCase().encode(
+          DigestUtils.getDigestWithManualFallbackWhenSizeUnknown(filePath, xattrProvider)
+        );
+      } else {
+        actualChecksum = getChecksum(keyType, filePath);
+      }
```

Comment:
looks like `Path` has a `getFastDigest` method that you can use here.

Also, we don't know what the digest algorithm is here, right? What if the `keyType` here is not the same as what xattr returns?

Reply:
> looks like Path has a getFastDigest method that you can use here.
`#getFastDigest` returns `null` if there's no digest. We need a digest here so then we'd need to case on it the response here. `#getDigest` recommends using `DigestUtils#getDigestWithManualFallback`.

> What if the `keyType` here is not the same as what xattr returns?
I think there's an implicit assumption today that the xattr needs to return the correct digest. I don't particularly love that assumption but I don't think it would be any worse to use it here vs. where it's currently being used to determine the checksum of other (non-repository-cache) files.

Reply:
> I think there's an implicit assumption today that the xattr needs to return the correct digest. I don't particularly love that assumption but I don't think it would be any worse to use it here vs. where it's currently being used to determine the checksum of other (non-repository-cache) files.

I don't get it. Where else is the xattr digest used? If it's used to compare against itself, it's obviously fine. But here we're saying "the xattr digest needs to match 'dcfa84b7...', which is the SHA-256 hash of this blob", OR "the xattr digest needs to match 'd98f13a...', which is the SHA-512 hash of this blob". What does it even mean for xattr to "return the correct digest"? There are multiple correct digests, depending on the `keyType`.

Reply:
> `Path#getFastDigest` returns `null` if there's no digest. We need a digest here so then we'd need to case on it the response here. `Path#getDigest` recommends using `DigestUtils#getDigestWithManualFallback`.

I think I'd rather we use `Path#getFastDigest`, and fall back to `getChecksum` if it returns null. The reason is that we don't get `XattrProvider` involved at all that way. The fact that `XattrProvider` makes no promises about the digest algorithm it uses makes me uneasy, and it doesn't help that we get the digest algorithm through a seemingly unrelated object (the `FileSystem`). If we use `Path#getFastDigest`, at least we can see that we're using the `FileSystem` of the `Path` to determine what digest algorithm we're using.

Reply:
I think that makes sense.

However, I'm concerned that `Path#getFastDigest` will in practice actually not have the correct digest algorithm associated with it.

Here's how one gets built:
1. https://github.com/sushain97/bazel/blob/c3e2d296f13f6c0c044a68035e912a2173656fa3/src/main/java/com/google/devtools/build/lib/bazel/repository/cache/RepositoryCache.java#L185
2. https://github.com/sushain97/bazel/blob/c3e2d296f13f6c0c044a68035e912a2173656fa3/src/main/java/com/google/devtools/build/lib/bazel/repository/cache/RepositoryCache.java#L182
3. https://github.com/sushain97/bazel/blob/c3e2d296f13f6c0c044a68035e912a2173656fa3/src/main/java/com/google/devtools/build/lib/bazel/repository/cac
4. https://github.com/sushain97/bazel/blob/c3e2d296f13f6c0c044a68035e912a2173656fa3/src/main/java/com/google/devtools/build/lib/bazel/repository/cache/RepositoryCache.java#L95he/RepositoryCache.java#L181
5. https://github.com/sushain97/bazel/blob/c3e2d296f13f6c0c044a68035e912a2173656fa3/src/main/java/com/google/devtools/build/lib/bazel/BazelRepositoryModule.java#L339
6. https://github.com/sushain97/bazel/blob/c3e2d296f13f6c0c044a68035e912a2173656fa3/src/main/java/com/google/devtools/build/lib/bazel/BazelRepositoryModule.java#L332
7. https://github.com/sushain97/bazel/blob/c3e2d296f13f6c0c044a68035e912a2173656fa3/src/main/java/com/google/devtools/build/lib/bazel/BazelRepositoryModule.java#L266
8. https://github.com/sushain97/bazel/blob/c3e2d296f13f6c0c044a68035e912a2173656fa3/src/main/java/com/google/devtools/build/lib/runtime/BlazeRuntime.java#L1273

OK, honestly, I'm losing the thread here - there's a lot of abstraction. But, even without digging into it further, I think the issue is that `filePath` is built from `contentAddressablePath` which is associated with a fixed filesystem and therefore a fixed digest that will not vary with `keyType`.

Reply:
> But, even without digging into it further, I think the issue is that `filePath` is built from `contentAddressablePath` which is associated with a fixed filesystem and therefore a fixed digest that will not vary with `keyType`.

That's okay -- I'm not asking the filesystem's digest function to vary with `keyType`. I just need to know that the fixed digest function that the filesystem is using is the same as `keyType`. If it doesn't match, we can always go and compute the hash from scratch.

It does sound like we're sort of working on a "trust" basis here, though -- the xattr digest is not necessarily computed with the `FileSystem#digestFunction` and could technically be injected by anyone. But since this whole PR is very much opt-in, I think it should be okay.

Reply:
Agreed. I've modified the PR accordingly - it is much smaller now. Appreciate you working through this with me.

Trying to find a way to write tests for the modified `getChecksum`...